R-packages/evalcast/R/get_predictions.R

```diff
@@ -69,79 +82,112 @@ get_predictions <- function(forecaster,
                             apply_corrections = function(signals) signals,
                             response_data_source = signals$data_source[1],
                             response_data_signal = signals$signal[1],
-                            forecaster_args = list()) {
-  
-  assert_that(is_tibble(signals), msg="`signals` should be a tibble.")
-  
-  
-  out <- forecast_dates %>%
-    map(~ do.call(
-          get_predictions_single_date,
-          list(forecaster = forecaster,
-               signals = signals,
-               forecast_date = .x,
-               apply_corrections = apply_corrections,
-               forecaster_args = forecaster_args)
-          )
-        ) %>%
-    bind_rows()
+                            forecaster_args = list(),
+                            parallel_execution = TRUE,
+                            additional_mclapply_args = list(),
+                            honest_as_of = TRUE,
+                            offline_signal_dir = NULL) {
+
+  assert_that(is_tibble(signals), msg = "`signals` should be a tibble.")
+  assert_that(xor(honest_as_of, "as_of" %in% names(signals)), msg = "`honest_as_of` should be set if and only if `as_of` is not set. Either remove as_of specification or set honest_as_of to FALSE.")
+  incidence_period <- match.arg(incidence_period)
+
+  if (incidence_period == "epiweek") rlang::warn("incidence_period of 'epiweek' only selects weekly (Saturday) target end dates, actual 'epiweek' signals not supported.", "evalcast::get_predictions")
+  if (!is.null(signals$time_type) & (signals$time_type == "week" || signals$time_type == "epiweek")) rlang::abort("time_type in the signals arg can only be 'day'.", "evalcast::get_predictions")
```

Comment:
tibble's generating some warnings in the exploration script: "Unknown or uninitialised column: `time_type`."

Background: `data.frame`'s `$` has some undesirable behaviors when programming:
```
data.frame(aaa=1:5)$aaa # 1:5, good
data.frame(aaa=1:5)$b # NULL, "convenient"
data.frame(aaa=1:5)$a # 1:5, "convenient"
data.frame(aaa=1:5, aab=2:6)$a # NULL, ..............
data.frame(time_type_NOT = c("a","b"))$time_type # "a" "b" rather than NULL; unlikely to come up with time_type but a concern for shorter names
```
Now, `tibble` tries to make this better behaved.  The partial matching behavior was not common(ly intended to be used in) code and brains, but  `is.null(df$col)` was.  I imagine this is why they chose to disable partial matching (giving NULL + warning instead), but instead of giving an error when requesting a column that does not exist, which would be natural if we ignore historical context, giving NULL + a warning as a compromise.  So how to do this without a warning?
```
!is.null(signals$time_type) # bad behavior for data.frame, okay but warning behavior for tibble
# -->
!is.null(signals[["time_type"]]) # expected behavior without warning for both
"time_type" %in% names(signals) # another way
rlang::has_name(signals, "time_type") # another way
```
the three ways here work and seem equivalent for data.frames, tibbles, as well as for lists provided that you aren't putting in "" instead of "time_type".

Short-circuiting scalar vs. non-short-circuiting vector logical operations:
```
c(FALSE, TRUE) || c(FALSE,TRUE) # FALSE, without even a warning!  (`if` directly will give a warning, but not `||` apparently)
any(c(FALSE, TRUE) | c(FALSE,TRUE)) # TRUE

rlang::has_name(signals, "time_type") & signals$time_type == "blah" # if no column, logical(0) + warning
rlang::has_name(signals, "time_type") & (signals$time_type == "blah" || signals$time_type == "blah2") # if no column, FALSE + 2 warnings
rlang::has_name(signals, "time_type") && signals$time_type == "blah" # if no column, right result, but if there is a column, result based on 1st entry only, no warning!
# vs say
rlang::has_name(signals, "time_type") && any(signals$time_type == "blah") # looking better
rlang::has_name(signals, "time_type") && any(signals$time_type == "blah" & ......) # looking better
```



This is intended to warn people away from usage that could have them trying to use a column that they thought would exist but doesn't, getting NULL, and then getting confusing errors or garbage results, OR, using a prefix that they thought would uniquely identify a column they expected to exist, but it doesn't, and getting NULL or some random column instead, OR, as here, checking whether a column exists when it doesn't but it matches a prefix exactly one other existing column, and getting false positive results.....  Lots can go wrong.
`!is.null(signals$time_type)` -> `!rlang::has_name(signals, "time_type")`

Reply:
(By the way, partial matching is also used for argument passing, without even a warning by default.  "Convenient".  This is probably one of the motivations for using dots as prefixes in dplyr function parameter names.  Something relevant to epiprocess and epipredict more than evalcast and the hospitalization forecaster, but keep in mind.)

Reply:
(Corrected some comment above about behavior being the same for lists... it isn't quite.)

Reply:
Thanks for writing this out, I will need to keep a link to this comment for future reference. Do you recommend any particular version that seems to avoid these issues? I'm thinking of just consistently using `"key" in names(df)`.

Reply:
Writing this out was also helpful for me.  A few more notes.
- I'm considering using [this advice](https://kevinushey.github.io/blog/2015/02/02/rprofile-essentials/) minus `options(warn=2)` (that's a nice option to use while debugging stuff though).
- Some of the short-circuiting silent gotchas surprised me.  From [here](https://stat.ethz.ch/R-manual/R-devel/library/base/html/Logic.html): "Using vectors of more than one element in && or || will give a warning (as from R 4.2.0), or an error if the environment variable _R_CHECK_LENGTH_1_LOGIC2_ is set to a true value (this is intended to become the default in future). ".  Don't think 4.2 is out in many OS's yet, though(?).  Hope at least `Sys.setenv("_R_CHECK_LENGTH_1_LOGIC2_"=TRUE)` might help catch some of these if it works on <4.2.
- Might want to research if there is a way to try to check these issues and make them into errors within the context of our package's operations.

Regarding the name-check: not sure, I might recommend with caveats what you suggested: `"key" %in% names(df)` sounds good for checking for column names we know at "compile time", and should also work for atomic vectors and lists as well; just don't try `"" %in% names(df)` or for unchecked user-input(-derived) things in place of "key" that could include "" --- whatever you want it to do / makes sense given `?names` docs etc., it's probably not going to consistently do it.  Since it's not completely robust/general for programming purposes, I'm not completely happy with it, but I'm also not completely happy with any of the other choices I listed above (or the additional other choice I just stumbled across, `%has_name%` from `assertthat`), for similar sorts of reasons.

Reply:
Nice finds. I'll try out `options(warn = 2)` for a while and see how it works.